src/dataflow/src/source/s3.rs

```diff
@@ -836,6 +817,60 @@ async fn download_object(
     }
 }
 
+async fn read_object_chunked<R>(
+    source_id: &str,
+    object_size: usize,
+    reader: &mut R,
+    tx: &Sender<Result<InternalMessage, S3Error>>,
+) -> (DownloadStatus, Option<DownloadMetricUpdate>)
+where
+    R: Unpin + AsyncRead,
+{
+    let mut bytes = 0;
+    let mut chunks = 0;
+    loop {
+        let mut chunk = vec![0; cmp::min(CHUNK_SIZE, object_size)];
+
+        let read_bytes = match reader.read(&mut chunk).await {
+            Ok(read_bytes) => read_bytes,
+            Err(err) => {
+                return (DownloadStatus::Fail(S3Error::Read(err)), None);
+            }
+        };
+
+        bytes += read_bytes;
+
+        if read_bytes > 0 {
+            chunks += 1;
+            chunk.truncate(read_bytes);
+            chunk.shrink_to_fit();
```

Comment:
Why's this necessary?

Reply:
It's not necessary, it just seemed reasonable to send the smallest amount of data into dataflow threads as possible.